zerver/views/message_fetch.py

```diff
@@ -591,6 +617,17 @@ def ok_to_include_history(narrow: OptionalNarrowListT, user_profile: UserProfile
     # query that narrows to a particular public stream on the user's realm.
     # If we screw this up, then we can get into a nasty situation of
     # polluting our narrow results with messages from other realms.
+
+    # For web-public queries, we are always returning history.  The
+    # analogues of the below stream access checks for whether streams
+    # have is_web_public set and banning is operators in this code
+    # path are done directly in NarrowBuilder.
+    if is_web_public_query:
+        assert user_profile is None
```

Comment:
A logged in user making a web-public query might be a strange thing to do, but do we want to ban it?

Reply:
When we do:
```python
    if not maybe_user_profile.is_authenticated:
        # If user is not authenticated, clients must include
        # `streams:web-public` in their narrow query to indicate this
        # is a web-public query.  This helps differentiate between
        # cases of web-public queries (where we should return the
        # web-public results only) and clients with buggy
        # authentication code (where we should return an auth error).
        if not is_web_public_narrow(narrow):
            return json_unauthorized()
        assert narrow is not None
        if not is_web_public_compatible(narrow):
            return json_unauthorized()

        realm = get_realm_from_request(request)
        if realm is None:
            return json_error(_("Invalid subdomain."))

        # We use None to indicate unauthenticated requests as it's more
        # readable than using AnonymousUser, and the lack of Django
        # stubs means that mypy can't check AnonymousUser well.
        user_profile: Optional[UserProfile] = None
        is_web_public_query = True
```
We ensure user_profile is None.

Reply:
User can still access the web-public streams in a normal query and we have tests for both these cases. So, we will know if anything weird is happening.

Reply:
Yeah, I think this is correct.